src/components/implementation/no_interface/scheddev/sl_timer_mod_periodic.c

```diff
@@ -1,6 +1,54 @@
 #include <sl.h>
 #include <sl_consts.h>
 #include <sl_plugins.h>
+#include <heap.h>
+
+#define SL_TIMEOUT_MOD_MAX_THDS MAX_NUM_THREADS
+#define SL_TIMEOUT_HEAP()       (&wakeup_heap.h)
+
+struct wakeup_heap {
+	struct heap  h;
+	void        *data[SL_TIMEOUT_MOD_MAX_THDS];
+} wakeup_heap;
+
+/* wakeup any blocked threads! */
+void
+sl_timeout_mod_wakeup_expired(cycles_t now)
+{
+	if (!heap_size(SL_TIMEOUT_HEAP())) return;
+
+	do {
+		struct sl_thd *tp, *th;
+
+		tp = heap_peek(SL_TIMEOUT_HEAP());
+		assert(tp);
+
+		/* FIXME: logic for wraparound in current tsc */
+		if (tp->wakeup_cycs > now) break;
+
+		th = heap_highest(SL_TIMEOUT_HEAP());
+		assert(th && th == tp);
+		th->wakeup_idx = -1;
+
+		sl_thd_wakeup_no_cs(th);
+	} while (heap_size(SL_TIMEOUT_HEAP()));
+}
+
+void
+sl_timeout_mod_block(struct sl_thd *t, cycles_t wakeup)
+{
+	assert(t && t->wakeup_idx == -1); /* not already in heap */
+	assert(heap_size(SL_TIMEOUT_HEAP()) < SL_TIMEOUT_MOD_MAX_THDS);
+
+	if (!wakeup) {
+		assert(t->period);
+		t->wakeup_cycs += t->period; /* implicit wakeup = task period */
```

Comment:
Yeah, I have a couple of `/*FIXME: wraparound*/`, I guess asserting here would help to address those comments eventually. I'll add that. Thanks.